src/inmanta/env.py

```diff
@@ -47,42 +52,236 @@
+                for requirement in requirements
+                if f"No matching distribution found for {requirement.project_name}" in stderr
+            ]
+            if not_found:
+                raise PackageNotFound("Packages %s were not found in the given indexes." % ", ".join(not_found))
+            raise e
+
+    def install_from_source(self, paths: List[LocalPackagePath]) -> None:
         """
         Install one or more packages from source. Any path arguments should be local paths to a package directory.
         """
         if len(paths) == 0:
             raise Exception("install_from_source requires at least one package to install")
-        # make sure we only try to install from a local source
         explicit_paths: Iterator[LocalPackagePath] = (
-            LocalPackagePath(path=os.path.join(".", path.path), editable=path.editable) for path in paths
+            # make sure we only try to install from a local source: add leading `./` and trailing `/` to explicitly tell pip
+            # we're pointing to a local directory.
+            LocalPackagePath(path=os.path.join(".", path.path, ""), editable=path.editable)
+            for path in paths
         )
-        subprocess.check_call(
+        self._run_command_and_log_output(
             [
-                cls.env_path,
+                self.python_path,
                 "-m",
                 "pip",
                 "install",
                 *chain.from_iterable(["-e", path.path] if path.editable else [path.path] for path in explicit_paths),
-            ]
+            ],
+            stderr=subprocess.PIPE,
         )
 
+    @classmethod
+    def _run_command_and_log_output(
+        cls, cmd: List[str], env: Optional[Dict[str, str]] = None, stderr: Optional[int] = None
+    ) -> str:
+        output: bytes = b""  # Make sure the var is always defined in the except bodies
+        try:
+            output = subprocess.check_output(cmd, stderr=stderr, env=env)
+        except CalledProcessError as e:
+            LOGGER.error("%s: %s", cmd, e.output.decode())
+            raise
+        except Exception:
+            LOGGER.error("%s: %s", cmd, output.decode())
+            raise
+        else:
+            LOGGER.debug("%s: %s", cmd, output.decode())
+            return output.decode()
+
+
+class ActiveEnv(Environment):
+    """
+    The active Python environment. Method implementations assume this environment is active when they're called.
+    Activating another environment that inherits from this one is allowed.
+    """
+
+    def install_from_index(
+        self, requirements: List[Requirement], index_urls: Optional[List[str]] = None, upgrade: bool = False
+    ) -> None:
+        super().install_from_index(requirements, index_urls, upgrade)
+        self.notify_change()
+
+    def install_from_source(self, paths: List[LocalPackagePath]) -> None:
+        super().install_from_source(paths)
+        self.notify_change()
```

Comment:
Because `notify_change` is only relevant to notify the active environment. It updates `pkg_resources.working_set`, `sys.site` and `importlib`, which affect the active environment only. I've recently pushed a commit that expanded on the `notify_change` constructor to illustrate its purpose in more detail, have you seen that one? Does it help clarify this?

Reply:
I agree.